packages/kbn-monaco/src/painless/worker/painless_worker.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { PainlessCompletionResult, PainlessContext, Field } from '../types';
+
+import { PainlessCompletionService } from './services';
+
+import {
+  isDeclaringField,
+  isConstructorInstance,
+  isAccessingProperty,
+  showStaticSuggestions,
+} from './utils';
+
+export class PainlessWorker {
+  async provideAutocompleteSuggestions(
+    currentLineChars: string,
+    context: PainlessContext,
+    fields?: Field[]
+  ): Promise<PainlessCompletionResult> {
+    const completionService = new PainlessCompletionService(context);
+    // Array of the active line words, e.g., [boolean, isTrue, =, true]
+    const words = currentLineChars.replace('\t', '').split(' ');
+    // What the user is currently typing
+    const activeTyping = words[words.length - 1];
+
+    const primitives = completionService.getPrimitives();
+
+    let autocompleteSuggestions: PainlessCompletionResult = {
+      isIncomplete: false,
+      suggestions: [],
+    };
+
+    if (isConstructorInstance(words)) {
```

Comment:
This is a slightly random thought I wanted to share mostly based on stylistic preference -- I think the current implementation is great! However:

I think it would be simpler to reduce the API of the `PainlessCompletionService` class to just `provideAutocompleteSuggestions` and move the if else logic here into that function - which maps to other functions. That way `getConstructorSuggestions` and others can become `private` methods and the constructor for the service takes in `currentLineChars` and `fields?` too. I feel like this could reduce indirection slightly (moving things closer together). Or, instead of a class, it could just be a function we call that returns the completion items.

Just wanted to hear your thoughts on this - no need to action! To my mind classes really shine when the state is more long lived and fully encapsulated.

Reply:
Good point! I think this is a great suggestion and will make the code easier to reason. I'll work on updating this.